backend/dao/winesDAO.js

```diff
@@ -1,3 +1,305 @@
+
+//todo: Take out console logs for submissions (clean it up!)
+//todo: figure out HOW to put these functions in the function.js file!
+
+//? ************************************* */
+//* ************Declarations************* */
+//* ************************************* */
+let wines;
+let query = {};
+let singleFilter = {};
+let filterObject = {};
+let filter_ = [];
+let price_type = [];
+let query1 = {};
```

Comment:
Can you walk through how this values interact with themselves in the DAO?
Well, after the explanation I'll most likely recommend some changes.

Reply:
Yes of course- I did comment a little bit in the code, but did not want to over comment- is this a situation that I should comment what everything represents?  
#### The reason I have these variables is to conform to the MongoDb query structures. 
The two major queries that need an alternative query structure are for **price** and **wine profile** (sweet, dry, bitter). 
query is the empty object that will be passed through the DB. After each queryFunction, the query object has a value added to it. For example, after the vegan function, a vegan property is added to the object with: `query.vegan = {query}.`
<br>
So, if one filter is added to the data object, it is simple to add the property to the current query. 
<br>
##### Docs: But, the problem that I ran into is that if a filter is a conditional, the format it this (example taken from docs) :
` { $or: [ { quantity: { $lt: 20 } }, { price: 10 } ] }`
**My next example will be using the price_eur scenario, but the same logic applies for the profile scenario**
The problem I ran into- this $or object property is a list, and I had to create an empty list so that, if for example, two prices ranges are sent through, the results will return both prices, instead of the last one that was passed through. Becuase without the list, at the last iteration of assigning filters through the query functions, the query would reassign to the last instance of the filter. 
*Example of bug:
`filters.price_eur = [med, low]`
*Goes through if statement to match query*
`query.price_eur = {medium prices}`
*second loop for low*
`query.price_eur = {low prices}`

*Instead what I need is to put both med and low into a list:
1. `singleFilter` is an object that will change with each iteration, but will be pushed to `filterObject.` 
ex: `singleFilter = {{ $gte: 0, $lt: 20.0 }}`
2. `filterObject` is a an instance to match the Mongodb query structure. filterObject could look like this: 
`filterObject = {price_eur : singleFilter} `
3. 'filter_' is a list instance to which filterObject pushes. This is to match the most outer list structure the MongoDB requires: `filter_ = [{price_eur : singleFilter}, {price_eur : singleFilter}]`
** with singleFilter having two different search queries.
4. query1 is another object to push the filter_ list to: it will look like this: which so far matches what the docs need, as stated above [docs](#docs)
`query1 = {'$or': [{price_eur : singleFilter}, {price_eur : singleFilter}] }`
5. I could have stopped here if there is not the possibility for another future conditional search, but there is: aka the profile type search. 
6. According to the docs, with a nested conditional, I need to match this datastructure: `{$and: [{$or: [{field: "value"}, {field: value}]}, {$or: [{"field": "value"}, {"field": value}]}]}`
7. So, `query1` from step 4 is added to the current empty list: `price_type`
8. In the wine profile (sweet, bitter, dry) function called: profileSelectors, the same proccess as above is implemented. 
9. Now, the final query is pushed to the` price_type` list to join the price queries. *Note to self, will rename price_type to more fitting name like: filter_list
10. Price_type list is added to the query object as:
`query.$and = price_type`
which, broken down, looks like:
`{$and: [{$or: [{price_eur: "value"}, {price_eur: value}]}, {$and: [{"profile": "value"}, {"profile": value}]}]}`

#### the above, last step, matches the mongodb docs for querying nested conditionals, and we are finally able to query any and all combinations of profile, price_eur, vegan, type, origin (maybe I am missing one..)

Reply:
Sorry, actually this level of details wasn't that necessary to my review.
There's some important things I'd like to explain in this review.
One is:
All the variables above are shared state between all the functions in this file.
A very important knowledge I want to share is: be very careful with shared state, and really try to avoid it unless it's completely necessary for the implementation, or it's a very small snippet of code that we can keep track of the state and it's very local to a function, and even in this case, the state changes of the shared state is sequential or easy to follow.
You should be specially careful when there's async code involved because then it's not sequential anymore, meaning the code can run in different order every time it's run in realtime.
What are the issues with shared state?
(one side note already, shared state is very common in OOP, and it's another thing that I hate about this paradigm)
1 - It's very hard to follow.
We have the variables defined on the top of the file, and in every function it is mutated or read, and when we read the code, we have no clear idea of what is the current state of the data, we need to jump around the file to understand it properly.
This will cause a lot of issues when someone tries to maintain or debug the code in the future.

2 - There's a lot of room for bugs.
In one function you are mutating the shared state and in other reading it.
Imagine that in one function you are mutating to be read in another part of the code by another function, then the code hits an async call, which means that it will be blocked and another code can run. Now imagine that this other code that runs is actually the call to the same function that mutates the shared state. When the first async call that was mentioned resolves, you may be expecting that the shared state is in a certain shape, but now it's completely different, and hard to reason about what shape it can be, and also hard to remember or be prepared to the fact that it can be different.

To wrap up, I think this file has a bug very easy to hit (or has a potential for bug pretty soon), but only happens when you have multiple requests to server at time, you can't hit this bug when testing alone.
The issues is mixing async code with share state.
Every time you have an `await` keyword in this file, the code will be blocked until it's resolved, the server can run another function of this file while it's awaiting though, so this means that the variable above can change anytime within any function that is async, meaning that you mutate the variables to a value, then another function call mutates it again to a different value, and you still haven't read the first value yet.
Does it make sense? It can be a bit hard to follow the logic with async code.
What would be the solution?
Actually pretty simple, use more parameters and local variables, meaning variables that exist only within the scope of the function.

There's also another point that you shouldn't mutate parameters passed to a function, which I'll explain more in another comment.